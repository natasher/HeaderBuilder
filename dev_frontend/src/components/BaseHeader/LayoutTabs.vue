<template>
  <div>

    <ul class = "small">

      <li v-for        = "( position, index ) in layoutPositions"
        :key           = "index"
        :data-position = "position"
        @click         = "handleLayoutPositionClick({ position: position })"
        :class         = "{
                            'active': ( getLayoutPosition === position ) ? true : false,
                          }"
        >
          <span class = "icon"></span>
      </li>

    </ul>

  </div>
</template>

<script>
import { mapActions, mapGetters } from 'vuex'

export default {

  name: 'LayoutTabs',

  data() {
    return {
      layoutPositions: [ 'top', 'bottom', 'left', 'right' ],
    }
  },

  computed: {
    ...mapGetters( 'ui', [
      'getLayoutPosition',
    ]),
  },

  methods: {
    ...mapActions( 'ui', [
      'handleLayoutPositionClick',
    ]),

  },

}
</script>
