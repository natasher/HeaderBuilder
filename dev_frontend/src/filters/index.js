/**
 * Capitalize first letter of str parameter.
 * e.g.
 *
 * "foo" => "Foo"
 * @param {string} str
 */
export const capitalize = ( value ) => {
  if ( !value ) return ''

  value = value.toString()
  return value.charAt( 0 ).toUpperCase() + value.slice( 1 )
}

/**
 * Convert string with dash separated words (e.g. value of data-type param)
 * to space separated string.
 * e.g.
 *
 * "menu_icon" => "menu icon"
 * @param {string} str
 */
export const convertDashToSpace = ( str ) => {
  if ( !str ) return ''

  return str.split( '_' ).join(' ')
}

/**
 * Convert string with space separated words to dash separated string.
 * e.g.
 *
 * "Action Bar" => "Action_Bar"
 * @param {string} str
 */
export const convertSpaceToDash = ( str ) => {
  if ( !str ) return ''

  return str.split(' ').join( '_' )
}

/**
 * Convert string with space to one Camel Cased string.
 * e.g.
 *
 * "background color" => "backgroundColor"
 * @param {string} str
 */
export const toCamelCase = ( str ) => {
  if ( !str ) return ''

  let arr    = str.split(' ')
  let output = arr.reduce(( prev, curr, index ) => {
    // return ( index === 0 ) ? prev + curr : prev + capitalize( curr )
    return (index === 0)
      ? curr.charAt(0).toLowerCase() + curr.slice(1)
      : prev + capitalize(curr)
  }, '')

  return output
}

/**
 * Convert camelCased string to human-readable string.
 * e.g.
 *
 * "backgroundColor" => "Background Color"
 * @param {string} str
 */
export const unCamelCase = ( str ) => {
  if ( !str ) return ''

  return str
    .replace(/([a-z])([A-Z])/g, '$1 $2')
    .replace(/\b([A-Z]+)([A-Z])([a-z])/, '$1 $2$3')
    .replace(/^./, ( str ) => { return str.toUpperCase() })
}

/**
 * Convert camelCased string to upper-cased snake case.
 * e.g.
 *
 * "secondRow" => "SECOND_ROW"
 */
export const allCaps = ( str ) => {
  if ( !str ) return ''

  return convertSpaceToDash(unCamelCase( str )).toUpperCase()
}