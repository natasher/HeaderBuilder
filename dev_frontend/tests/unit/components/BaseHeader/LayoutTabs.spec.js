import { shallowMount, createLocalVue } from '@vue/test-utils'
import { capitalize } from '../../../../src/filters/index'
import Vuex          from 'vuex'
import flushPromises from 'flush-promises'
import merge         from 'lodash.merge'
import LayoutTabs    from '../../../../src/components/BaseHeader/LayoutTabs.vue'

const localVue = createLocalVue()
localVue.use( Vuex )

function createStore(overrides) {
  const defaultStoreConfig = {
    modules: {
      ui: {
        namespaced: true,
        getters: {
          getLayoutPosition: jest.fn().mockReturnValue('top'),
        },
        actions: {
          handleLayoutPositionClick: jest.fn()
        },
      },
    }
  }

  return new Vuex.Store(
    merge( defaultStoreConfig, overrides )
  )
}

function createWrapper(overrides) {
  const defaultMountingOptions = {
    localVue,
    store: createStore(),
    filters: { capitalize },
  }

  return shallowMount( LayoutTabs, merge( defaultMountingOptions, overrides ) )
}

describe('LayoutTabs.vue', () => {

  describe('snapshots:', () => {

    test('should match if layout position is `top`', () => {
      const wrapper = createWrapper()

      expect( wrapper.element ).toMatchSnapshot()
    })

    test('should match if layout position is `bottom`', async () => {
      const store = createStore({
        modules: {
          ui: {
            getters: {
              getLayoutPosition: jest.fn().mockReturnValue('bottom'),
            },
          },
        }
      })
      const wrapper = createWrapper({ store, })

      expect( wrapper.element ).toMatchSnapshot()
    })

    test('should match if layout position is `left`', async () => {
      const store = createStore({
        modules: {
          ui: {
            getters: {
              getLayoutPosition: jest.fn().mockReturnValue('left'),
            },
          },
        }
      })
      const wrapper = createWrapper({ store, })

      expect( wrapper.element ).toMatchSnapshot()
    })

    test('should match if layout position is `right`', async () => {
      const store = createStore({
        modules: {
          ui: {
            getters: {
              getLayoutPosition: jest.fn().mockReturnValue('right'),
            },
          },
        }
      })
      const wrapper = createWrapper({ store, })

      expect( wrapper.element ).toMatchSnapshot()
    })


  })

  test('should add `.active` class to currently displayed layout', async () => {
    const pos = 'left'
    const store = createStore({
      modules: {
        ui: {
          getters: {
            getLayoutPosition: jest.fn().mockReturnValue( pos ),
          },
        },
      }
    })
    const wrapper = createWrapper({ store, })

    await flushPromises()

    expect( wrapper.find(`li[data-position="${ pos }"]`).classes() ).toContain( 'active' )
  })

  test('should invoke layout position when desired position is clicked', async () => {
    const handleLayoutPositionClick = jest.fn()
    const store = createStore({
      modules: {
        ui: {
          actions: {
            handleLayoutPositionClick,
          }
        }
      }
    })
    const wrapper = createWrapper({ store, })

    const pos     = wrapper.findAll('li').at( 3 )
    const posName = pos.attributes()[ 'data-position' ]

    pos.trigger('click')
    await flushPromises()

    expect( handleLayoutPositionClick ).toBeCalledWith( expect.anything(), { position: posName }, void 0)
  })

})