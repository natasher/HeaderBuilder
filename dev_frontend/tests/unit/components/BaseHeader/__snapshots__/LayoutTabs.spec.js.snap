// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LayoutTabs.vue snapshots: should match if layout position is \`bottom\` 1`] = `
<div>
  <ul
    class="small"
  >
    <li
      class=""
      data-position="top"
    >
      <span
        class="icon"
      />
    </li>
    <li
      class="active"
      data-position="bottom"
    >
      <span
        class="icon"
      />
    </li>
    <li
      class=""
      data-position="left"
    >
      <span
        class="icon"
      />
    </li>
    <li
      class=""
      data-position="right"
    >
      <span
        class="icon"
      />
    </li>
  </ul>
</div>
`;

exports[`LayoutTabs.vue snapshots: should match if layout position is \`left\` 1`] = `
<div>
  <ul
    class="small"
  >
    <li
      class=""
      data-position="top"
    >
      <span
        class="icon"
      />
    </li>
    <li
      class=""
      data-position="bottom"
    >
      <span
        class="icon"
      />
    </li>
    <li
      class="active"
      data-position="left"
    >
      <span
        class="icon"
      />
    </li>
    <li
      class=""
      data-position="right"
    >
      <span
        class="icon"
      />
    </li>
  </ul>
</div>
`;

exports[`LayoutTabs.vue snapshots: should match if layout position is \`right\` 1`] = `
<div>
  <ul
    class="small"
  >
    <li
      class=""
      data-position="top"
    >
      <span
        class="icon"
      />
    </li>
    <li
      class=""
      data-position="bottom"
    >
      <span
        class="icon"
      />
    </li>
    <li
      class=""
      data-position="left"
    >
      <span
        class="icon"
      />
    </li>
    <li
      class="active"
      data-position="right"
    >
      <span
        class="icon"
      />
    </li>
  </ul>
</div>
`;

exports[`LayoutTabs.vue snapshots: should match if layout position is \`top\` 1`] = `
<div>
  <ul
    class="small"
  >
    <li
      class="active"
      data-position="top"
    >
      <span
        class="icon"
      />
    </li>
    <li
      class=""
      data-position="bottom"
    >
      <span
        class="icon"
      />
    </li>
    <li
      class=""
      data-position="left"
    >
      <span
        class="icon"
      />
    </li>
    <li
      class=""
      data-position="right"
    >
      <span
        class="icon"
      />
    </li>
  </ul>
</div>
`;
